community/kernel/src/main/java/org/neo4j/kernel/impl/store/NeoStore.java

```diff
@@ -365,62 +363,78 @@ public int getRecordSize()
+    public static long setRecord( PageCache pageCache, File neoStore, Position
+            position, long
+            value )
             throws IOException
     {
-        int trailerSize = UTF8.encode( buildTypeDescriptorAndVersion( TYPE_DESCRIPTOR ) ).length;
-        try ( StoreChannel channel = fileSystem.open( neoStore, "rw" ) )
+        long previousValue = FIELD_NOT_INITIALIZED;
+        try ( PagedFile pagedFile = pageCache.map( neoStore, getPageSize( pageCache ) ) )
         {
-            long previous = FIELD_NOT_INITIALIZED;
-
-            long trailerOffset = channel.size() - trailerSize;
-            ByteBuffer buffer = ByteBuffer.allocate( RECORD_SIZE );
-            ByteBuffer trailerBuffer = null;
-            int recordOffset = RECORD_SIZE * position.id;
-            if ( recordOffset < trailerOffset )
-            {
-                // We're overwriting a record, get the previous value
-                channel.position( recordOffset );
-                channel.read( buffer );
-                buffer.flip();
-                buffer.get(); // inUse
-                previous = buffer.getLong();
-            }
-            else
-            {
-                // We're adding a new record, first cut off and keep the trailer
-                trailerBuffer = ByteBuffer.allocate( trailerSize );
-                channel.position( trailerOffset );
-                channel.read( trailerBuffer );
-                trailerBuffer.flip();
-                channel.truncate( trailerOffset );
-            }
-            buffer.clear();
-
-            // Write the value
-            channel.position( recordOffset );
-            buffer.put( Record.IN_USE.byteValue() );
-            buffer.putLong( value );
-            buffer.flip();
-            channel.write( buffer );
-
-            // Append the trailer if we cut it off previously
-            int newTrailerOffset = recordOffset + RECORD_SIZE;
-            if ( newTrailerOffset > trailerOffset )
+            if ( pagedFile.getLastPageId() != -1 )
             {
-                assert trailerBuffer != null;
-                channel.position( newTrailerOffset );
-                channel.write( trailerBuffer );
+                try ( PageCursor pageCursor = pagedFile.io( 0, PagedFile.PF_EXCLUSIVE_LOCK ) )
+                {
+                    if ( pageCursor.next() )
+                    {
+                        byte[] trailerData = UTF8.encode( buildTypeDescriptorAndVersion( TYPE_DESCRIPTOR ) );
+                        byte[] allData = new byte[pagedFile.pageSize()];
+                        do
+                        {
+                            pageCursor.getBytes( allData );
+                        }
+                        while ( pageCursor.shouldRetry() );
+                        int trailerOffset = StoreVersionCheck.findTrailerPattern( allData, trailerData );
```

Comment:
What do you think about renaming `StoreVersionCheck.findTrailerPattern` to `StoreVersionCheck.findTrailerOffset` instead?